cli.js

```diff
@@ -112,17 +117,32 @@ api.on('error', logger.unhandledError);
 api.on('stdout', logger.stdout);
 api.on('stderr', logger.stderr);
 
-api.run()
-	.then(function () {
-		logger.finish();
-		logger.exit(api.failCount > 0 || api.rejectionCount > 0 || api.exceptionCount > 0 ? 1 : 0);
-	})
-	.catch(function (err) {
-		if (err.name === 'AvaError') {
-			console.log('  ' + colors.error(figures.cross) + ' ' + err.message);
-		} else {
-			console.error(colors.stack(err.stack));
-		}
-
-		logger.exit(1);
+function run() {
+	api.run()
+		.then(function () {
+			logger.finish();
+			logger.exit(api.failCount > 0 || api.rejectionCount > 0 || api.exceptionCount > 0 ? 1 : 0);
+		})
+		.catch(function (err) {
+			if (err.name === 'AvaError') {
+				console.log('  ' + colors.error(figures.cross) + ' ' + err.message);
+			} else {
+				console.error(colors.stack(err.stack));
+			}
+
+			logger.exit(1);
 	});
+}
+
+run();
+
+var cwd = process.cwd();
+var watcher = chokidar.watch(cwd, {
+	persistent: cli.flags.watch,
+	ignored: [cwd + '/node_modules', cwd + '/.git'],
+	followSymlinks: false
+});
+
+watcher.on('change', function (filepath) {
+	run();
```

Comment:
Since the tests run asynchronously, you may end up starting a new batch before the previous execution has finished. It doesn't look like `api.run()` has any logic to defend against that (which it wouldn't have needed before).

Also `filepath` is unused.